wandb/sdk/launch/agent/agent.py

```diff
@@ -102,51 +137,42 @@ def print_status(self) -> None:
         output_str = "agent "
         if self._name:
             output_str += f"{self._name} "
-        if self._running < self._max_jobs:
+        if len(self._jobs) < self._max_jobs:
             output_str += "polling on "
             if self._project != LAUNCH_DEFAULT_PROJECT:
                 output_str += f"project {self._project}, "
             output_str += f"queues {','.join(self._queues)}, "
         output_str += (
-            f"running {self._running} out of a maximum of {self._max_jobs} jobs"
+            f"running {len(self._jobs)} out of a maximum of {self._max_jobs} jobs"
         )
 
         wandb.termlog(f"{LOG_PREFIX}{output_str}")
-        if self._running > 0:
-            output_str += f": {','.join([str(key) for key in self._jobs.keys()])}"
+        if len(self._jobs) > 0:
+            output_str += f": {','.join(str(job_id) for job_id in self.job_ids)}"
         _logger.info(output_str)
 
     def update_status(self, status: str) -> None:
         update_ret = self._api.update_launch_agent_status(
             self._id, status, self.gorilla_supports_agents
         )
         if not update_ret["success"]:
-            wandb.termerror(f"Failed to update agent status to {status}")
+            wandb.termerror(f"{LOG_PREFIX}Failed to update agent status to {status}")
 
-    def finish_job_id(self, job_id: Union[str, int]) -> None:
+    def finish_thread_id(self, thread_id: int) -> None:
         """Removes the job from our list for now."""
         # TODO:  keep logs or something for the finished jobs
-        del self._jobs[job_id]
-        self._running -= 1
+        with self._jobs_lock:
+            del self._jobs[thread_id]
         # update status back to polling if no jobs are running
-        if self._running == 0:
+        if len(self._jobs) == 0:
             self.update_status(AGENT_POLLING)
 
-    def _update_finished(self, job_id: Union[int, str]) -> None:
+    def _update_finished(self, thread_id: int) -> None:
         """Check our status enum."""
-        try:
-            if self._jobs[job_id].get_status().state in ["failed", "finished"]:
-                self.finish_job_id(job_id)
-        except Exception as e:
-            if isinstance(e, LaunchError):
-                wandb.termerror(f"Terminating job {job_id} because it failed to start:")
-                wandb.termerror(str(e))
-            _logger.info("---")
-            _logger.info("Caught exception while getting status.")
-            _logger.info(f"Job ID: {job_id}")
-            _logger.info(traceback.format_exc())
-            _logger.info("---")
-            self.finish_job_id(job_id)
+        with self._jobs_lock:
+            job = self._jobs[thread_id]
+        if job.job_completed():
```

Comment:
how do we make this an attr? like `@attr` or something so its not a function...

Reply:
@property